### PR TITLE
Increase Infer Sharing Abilities 

### PR DESCRIPTION
**Pull request by calebmkim**
There was a very specific scenario that was limiting our ability to infer sharing: 
```
group let25<"static"=2> {
      seq_mem.read_en = 1'd1; // 
      seq_mem.addr3 = ...; 
      seq_mem.addr2 = ...;
      seq_memaddr1 = ...;
      seq_mem.addr0 = ... ;
      reg.in = seq_mem.read_data;
      reg.write_en = seq_mem.read_done;
      let25[done] = reg.done;
}
```
We need this to count as a write to `reg`, so I created a function checking for this very specific case. 

(Also, the reason why I changed the pass order is the following: if we run the `compile-ref` pass, then it will turn `seq_mem.read_done` into a port `seq_mem_read_done` which no longer has the "done" attribute.  In other words, we need the `infer-share` pass to have information about the go-done ports of ref-cells. ).

### CODE REVIEW

File: calyx/src/analysis/domination_analysis/node_analysis.rs [hunk 1]
```diff
@@ -170,15 +170,73 @@ impl NodeSearch {
         NodeSearch { name }
     }
 
+    // 1) checks whether assign is of form dst.go = src.done;
+    // 2) checks whether assignments contains both:
+    // src.go = 1;
+    // group[done] = dst.done;
+    // If all of this is true, then returns true: we can be confident dst is
+    // written to. otherwise return false
+    fn special_case(
```
Comment (rachitnigam) [inline calyx/src/analysis/domination_analysis/node_analysis.rs:179]: More informative name needed here. Also, the comment doesn't define the relationship between `src` and `dst`

**Comment by rachitnigam** [comment]
I would've expected that this group would get split up by `group_to_seq` because the `out` from a `seq_mem` is stable so all this group is doing is reading the value from a memory and writing it into the register in the next cycle.

**Comment by calebmkim** [comment]
Sorry, I think my comment oversimplified the case a bit too much. `group2seq` doesn't handle the case when there are combinational components being written to in the group. So the example really has: 
```
slice.in = ... 
seq_mem.addr0 = slice.out; 
```
The reason why we didn't handle this originally is because we didn't want to have to deal with the case when (for example) the same slice is read from both cells (to connect it back to the given example, we didn't want to deal with the case for when `slice.out` was read from for both `seq_mem` and `reg`).

**Comment by rachitnigam** [comment]
Yeah, might be worth trying it out. One possible silly solution is cloning all the assignments to the combinational components in both the groups. Since we know the groups run in sequence, there will be no conflicts from doing. The one downside is that I'm not sure if we'll remove the unnecessary assignments from groups that don't use them

**Comment by rachitnigam** [comment]
Just checked–the `post-opt` optimizations do remove unused assignments from groups

**Comment by calebmkim** [comment]
Closing, covered by #1388

**Pull request closed without merging**